doc/release-notes.md

```diff
@@ -42,6 +42,18 @@ heights prior to this will not have any information recorded. To track changes
 from genesis, and thus monitor the total transparent pool size and chain supply,
 you would need to restart your node with the `-reindex` option.
 
+Wallet Performance Fixes
+------------------------
+
+The 100MiB memory limit for the batch scanner has been replaced by a 1000-block
+limit. This eliminates an expensive call to determine the current memory usage 
+of the batch scanner. 
```

Comment:
The minimum size per output is the minimum for Sprout ( $1698 / 2$ bytes, because there are two outputs in a JoinSplit), Sapling (948 bytes in either v4 or v5 transactions), or Orchard ( $820+2272+64 = 3156$ bytes), i.e. 948 bytes. So the worst-case number of outputs per block is $\mathsf{floor}\(\frac{2000000}{948}\) = 2109$, hence 2109000 outputs for 1000 blocks.
[TBC; also calculate the memory usage.]

Reply:
The memory usage of a batch of $2109000$ outputs will be at least the size of those outputs, i.e. ~2 GB. So I think that this change potentially introduces a significant memory regression.

Reply:
It won't generally be 2 GiB peak though, because the batch scanner flushes batches when they reach 20 outputs. So the largest single batch would be 2128 outputs (if transactions align perfectly), and then that will start trial decrypting before the next batch is created. The peak will definitely be higher than 100 MiB, but under reindexing conditions I think 2 GiB is also fine (and still bounded, unlike before we added the memory limiting).

Reply:
To clarify the numbers (now that I'm not typing from my phone):

- The wallet notification thread operates on a per-block level.
  - This is due to how the chain works. It notifies to the same granularity as the main node thread previously did.
- We don't know how many outputs are in the block until we are about to notify the wallets of its connection.
  - If we wanted to know, we'd either have to read every block twice from disk (doubling disk I/O pressure), or we'd have to store all the blocks in memory at the same time as we are trying to add outputs to the batch scanner (likely doubling the memory pressure).
- The batch scanner collects outputs into batches at a per-transaction level.
  - Doing per-output is in theory possible but would make the batch scanner logic significantly more complex as the usage side would need to join on a bunch of individual channels instead of a single channel.

This means that the largest a batch can be is a series of transactions that have a total of 19 outputs across them, followed by a single transaction that is as large as can fit into a single block, i.e. 2128 outputs. This would mean however that the block before must not have been filled with a single transaction, so the batch containing the rest of that block's transactions would have necessarily been smaller (no more than 2090 outputs). Alternatively, if the chain contained a sequence of blocks that each had a single max-size transaction, then you'd get a sequence of batches of 2109 outputs (as each transaction would on its own exceed the 20 output limit).

Reply:
The other component of this is "why 1000". The concern here is that if this number is too low, it artificially throttles the processing of small blocks on faster nodes, recreating the "wallet notification gets farther and farther behind the chain" problem that we are trying to solve. More specifically, if the node is able to notify X blocks in under 1 second, it has to spend the rest of that second waiting until the next second-boundary to get the next X blocks.

In a pairing, @nuttycom and I looked at how quickly his node (Ryzen 9 5950X, chain stored on an SSD) was reindexing the _chain_ (i.e. how quickly it was generating data that wallets needed to be notified of). During the pre-sandblasting region, it was processing around 250 blocks per second, and around half that before Blossom. This indicates that the bottleneck here was "chain bandwidth" (which wasn't altered much by the doubling of block throughput at Blossom).

Given this, we decided to set the limit to 1000 (4x the observed chain throughput), to give some headroom for even more powerful nodes, while limiting the worst-case likely batch scanner memory usage to 2 GiB.